pkg/checkpoint/process_test.go

```diff
@@ -272,12 +319,24 @@ func TestProcess(t *testing.T) {
 		if tc.podName != "" {
 			cp.PodName = tc.podName
 		}
-		gotStart, gotStop, gotRemove := process(tc.localRunning, tc.localParents, tc.apiParents, tc.activeCheckpoints, tc.inactiveCheckpoints, cp)
+		c := checkpoints{}
+
+		// Run test now.
+		now := time.Time{}
+		c.update(tc.localRunning, tc.localParents, tc.apiParents, tc.activeCheckpoints, tc.inactiveCheckpoints, cp)
+		gotStart, gotStop, gotRemove := c.process(now, tc.apiParents != nil, tc.localRunning, tc.localParents, tc.apiParents)
+
+		// Advance past grace period and test again.
+		now = now.Add(checkpointGracePeriod)
+		c.update(tc.localRunning, tc.localParents, tc.apiParents, tc.activeCheckpoints, tc.inactiveCheckpoints, cp)
+		gotGraceStart, gotGraceStop, gotGraceRemove := c.process(now, tc.apiParents != nil, tc.localRunning, tc.localParents, tc.apiParents)
 		if !reflect.DeepEqual(tc.expectStart, gotStart) ||
 			!reflect.DeepEqual(tc.expectStop, gotStop) ||
-			!reflect.DeepEqual(tc.expectRemove, gotRemove) {
-			t.Errorf("For test: %s\nExpected start: %s Got: %s\nExpected stop: %s Got: %s\nExpected remove: %s Got: %s\n",
-				tc.desc, tc.expectStart, gotStart, tc.expectStop, gotStop, tc.expectRemove, gotRemove)
+			!reflect.DeepEqual(tc.expectRemove, gotRemove) ||
+			!reflect.DeepEqual(tc.expectGraceStart, gotGraceStart) ||
```

Comment:
nit: Maybe just print them one by one? I mean:
```go
if !reflect.DeepEqual() {
    t.Errorf()
}
if !reflect.DeepEqual() {
    t.Errorf()
}
...
```

Reply:
If there are multiple states this can get a little verbose. I prefer to keep it as is.